rpcs3/Emu/system_progress.cpp

```diff
@@ -117,15 +117,27 @@ void progress_dialog_server::operator()()
 		u32 pdone  = 0;
 		auto text1 = text0;
 
+		const u64 start_time = get_system_time();
+
 		// Update progress
 		while (!g_system_progress_stopping && thread_ctrl::state() != thread_state::aborting)
 		{
-			const auto text_new = +g_progr.load();
+			auto whole_state = std::make_tuple(+g_progr.load(), +g_progr_ftotal, +g_progr_fdone, +g_progr_ptotal, +g_progr_pdone);
+
+			while (true)
+			{
+				const auto new_state = std::make_tuple(+g_progr.load(), +g_progr_ftotal, +g_progr_fdone, +g_progr_ptotal, +g_progr_pdone);
+
+				if (new_state == whole_state)
```

Comment:
And why should that make any difference?
It may change a nanosecond later.

Reply:
Before (race example in steps):
1. Progress dialog loads nullptr text.
2. Progress dialog loads ptotal. Let's say it is equal to X. pdone is equal to X as well here.
3. A PPU thread is about to compile a module, sets text and incrementes both ptotal and pdone.
4. Progress dialog loads pdone, pdone value in thread is now eqaul to X + 1, ptotal value in thread is X. In reality they are both X + 1.
5. Progress dialog exists the loop, resets them all. now pdone is ever lagging by 1.

After:
1. Progress dialog loads nullptr text.
2. Progress dialog loads ptotal. Let's say it is equal to X. pdone is equal to X here as well here.
3. A PPU thread is about to compile a module, sets text and incremenetes both ptotal and pdone.
4. Progress dialog loads pdone, pdone value in thread is now eqaul to X + 1, ptotal value in thread is X. In reality they are both X + 1.
5. Progress dialog sees value mismatch, restarts loads and gets equal ptotal and pdone and does not exit the loop.

Another example:
Before:
1. Progress dialog loads nullptr text.
2. A PPU thread is about to compile a module, sets text and incrementes both ptotal and pdone.
3. Progress dialog loads pdone, pdone value in thread is now eqaul to X + 1, ptotal value in thread is X + 1. In reality they are both X + 1.
4. Progress dialog loads ptotal. Let's say it is equal to X + 1. pdone is equal to X as well here.
5. Progress dialog exists the loop, resets them all. now both pdone and ptotal are ever lagging by 1.

After:
1. Progress dialog loads nullptr text.
2. A PPU thread is about to compile a module, sets text and incrementes both ptotal and pdone.
3. Progress dialog loads pdone, pdone value in thread is now eqaul to X + 1, ptotal value in thread is X + 1. In reality they are both X + 1.
4. Progress dialog loads ptotal. Let's say it is equal to X + 1. pdone is equal to X as well here.
5. Progress dialog sees value mismatch, restarts loads and gets equal ptotal and pdone and does not exit the loop.


TLDR: Thread-safety is a headache.

Reply:
That's purely random now, same as before, if this loop runs faster than the next change is supplied.
There's zero thread safety here.

You should've just used a struct with a mutex and call it a day.

Reply:
No there is no race here now. The fact that SPU does not crash for games is a live example for it. GETLLAR is an atomic 128-byte load. This loop recheck is a technique to load atomicily any amount of bytes without a mutex,